pkg/cloudprovider/providers/aws/aws_loadbalancer.go

```diff
@@ -308,3 +310,82 @@ func (s *AWSCloud) ensureLoadBalancerInstances(loadBalancerName string, lbInstan
 
 	return nil
 }
+
+func getLoadBalancerHostedZone(service *api.Service) string {
+	if zone, ok := service.Annotations[ServiceAnnotationLoadBalancerCnameZone]; ok {
+		return zone + "."
+	} else {
+		return ""
+	}
+}
+
+func (s *AWSCloud) getLoadBalancerCname(service *api.Service) string {
+	hostedZone := getLoadBalancerHostedZone(service)
+	if hostedZone == "" {
+		return ""
+	}
+
+	return fmt.Sprintf("%v-%v-%v.%v", service.Name, service.Namespace, s.cfg.Global.KubernetesClusterTag, hostedZone)
+}
+
+func (s *AWSCloud) updateLoadBalancerCnameEntry(service *api.Service, dnsName *string, action string) error {
+	hostedZone := getLoadBalancerHostedZone(service)
+	if hostedZone == "" {
+		return nil
+	}
+
+	s.r53Mutex.Lock()
+	defer s.r53Mutex.Unlock()
+
+	zonesInput := &route53.ListHostedZonesByNameInput{DNSName: aws.String(hostedZone)}
+	zones, err := s.r53.ListHostedZonesByName(zonesInput)
+	if err != nil {
+		return err
+	}
+
+	if len(zones.HostedZones) == 0 {
+		return fmt.Errorf("No hosted zones found for account")
```

Comment:
This could happen if hostedZone > existing zones, right?

Might be better to simply the logic into

`if len(zones.HostedZones) == 0 || aws.StringValue(zones.HostsZones[0].Name) != hostedZone`